src/io/fs_store.rs

```diff
@@ -52,28 +52,84 @@ impl FilesystemStore {
 
 impl KVStore for FilesystemStore {
 	type Reader = FilesystemReader;
-	type Writer = FilesystemWriter;
 
 	fn read(&self, namespace: &str, key: &str) -> std::io::Result<Self::Reader> {
 		let mut outer_lock = self.locks.lock().unwrap();
 		let lock_key = (namespace.to_string(), key.to_string());
 		let inner_lock_ref = Arc::clone(&outer_lock.entry(lock_key).or_default());
 
-		let mut dest_file = self.dest_dir.clone();
-		dest_file.push(namespace);
-		dest_file.push(key);
-		FilesystemReader::new(dest_file, inner_lock_ref)
+		let mut dest_file_path = self.dest_dir.clone();
+		dest_file_path.push(namespace);
+		dest_file_path.push(key);
+		FilesystemReader::new(dest_file_path, inner_lock_ref)
 	}
 
-	fn write(&self, namespace: &str, key: &str) -> std::io::Result<Self::Writer> {
+	fn write(&self, namespace: &str, key: &str, buf: &[u8]) -> std::io::Result<()> {
 		let mut outer_lock = self.locks.lock().unwrap();
 		let lock_key = (namespace.to_string(), key.to_string());
 		let inner_lock_ref = Arc::clone(&outer_lock.entry(lock_key).or_default());
+		let _guard = inner_lock_ref.write().unwrap();
+
+		let mut dest_file_path = self.dest_dir.clone();
+		dest_file_path.push(namespace);
+		dest_file_path.push(key);
+
+		let msg = format!("Could not retrieve parent directory of {}.", dest_file_path.display());
+		let parent_directory = dest_file_path
+			.parent()
+			.ok_or(std::io::Error::new(std::io::ErrorKind::InvalidInput, msg))?
```

Comment:
Unrelated to this PR, but may want to use `ok_or_else` here and elsewhere to avoid allocating a string unnecessarily.